### PR TITLE
Fix ActiveDocumentTracker bugs

### LINKED ISSUE #2079: Taint filtering does not work for single-click navigation in the Solution Explorer
### Description

The taint issues list is supposed to be filtered to only show issues for the active document. It works if the user navigates by double-clicking in the Solution Explorer, but not if the VS is configured to navigate by single-clicking in the Solution Explorer.

Note: this bug will also affect the shipped pre-compiled headers code. However, that won't be as visible to the user as it is a performance optimisation and there is no visual indication that anything is wrong. The CFamily issues will still be calculated correctly, but won't benefit from the optimisation.

Related: #2091

### Repro steps

1. Enable single-click navigation in the Solution Explorer by allowing files to be previewed in the Error List:

image

2. Open a bound solution contain taint issues.
3. Navigate between files by single-clicking in the Solution Explorer.

### Expected behavior

The taint list is filtered to the previewed document.

### Actual behavior

The filter is not updated.

### Known workarounds

Disable single-click preview / navigate by double-clicking

### Related information

* SonarLint for VS Version: v4.32

### Additional notes
The affected component is `ActiveDocumentTracker`. It uses `IVsSelectionEvents` to detect when a new file has selection. In the case of single-click navigation, the file will be opened but the selection remains in the Solution Explorer.

### LINKED ISSUE #2091: Taint filtering does not work when the last document is closed
### Description

The taint tool window caption is not updated correctly when the last document in a bound solution is closed.
The same issue can occur when switching from a bound project to an unbound project that doesn't have any open files.

Related: #2079.

Like #2079, the problem is in the underlying `ActiveDocumentTracker`, which doesn't raise an event when the last document is closed. If the event were raised then the taint caption would be updated correctly.

The issue is cosmetic.

### Repro steps

1. Open a solution with taint issues.
2. Close all files exception one that contains taint issues (=> should have data in the taint list)
3. Close that file

### Expected behavior

The taint tool window caption reset to the default and the taint list should be cleared.

### Actual behavior

The caption is not updated.
 
image

### Known workarounds

Close the solution / open another file

### Related information

* SonarLint for VS Version: 4.32

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixes #2079
Fixes #2091

### CODE REVIEW

File: src/Infrastructure.VS.UnitTests/ActiveDocumentTrackerTests.cs [hunk 1]
```diff
@@ -91,12 +93,12 @@ public void Dispose_ShouldUnregisterFromSelectionEvents()
         [TestMethod]
         public void Dispose_ShouldNoLongerRaiseEvents()
         {
-            var selectedFrame = CreateMockFrame(__WindowFrameTypeFlags.WINDOWFRAMETYPE_Document);
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS.UnitTests/ActiveDocumentTrackerTests.cs:94]: The product code doesn't need to check the frame type property any more; we are only interested in SEID_DocumentFrame notifications, so we know that the frame must be a document frame,

File: src/Infrastructure.VS.UnitTests/ActiveDocumentTrackerTests.cs [hunk 2]
```diff
@@ -91,12 +93,12 @@ public void Dispose_ShouldUnregisterFromSelectionEvents()
...
+            var selectedFrame = ValidWindowFrame;
 
             testSubject.Dispose();
-            (testSubject as IVsSelectionEvents).OnElementValueChanged((uint)VSConstants.VSSELELEMID.SEID_WindowFrame, "1", selectedFrame.Object);
+            (testSubject as IVsSelectionEvents).OnElementValueChanged((uint)VSConstants.VSSELELEMID.SEID_WindowFrame, "1", selectedFrame);
 
-            mockEventHandler.Verify(x => x(It.IsAny<ActiveDocumentChangedEventArgs>()), Times.Never);
+            CheckEventNotRaised();
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS.UnitTests/ActiveDocumentTrackerTests.cs:101]: Unrelated - added helper methods to make the tests easier to read.

File: src/Infrastructure.VS.UnitTests/ActiveDocumentTrackerTests.cs [hunk 3]
```diff
@@ -114,115 +116,85 @@ public void OnSelectionChanged_DoesNotRaiseEvent()
...
-        public void OnElementValueChanged_NonWindowFrameElement_DoesNotRaiseEvent()
+        [DataRow(VSConstants.VSSELELEMID.SEID_PropertyBrowserSID)]
+        [DataRow(VSConstants.VSSELELEMID.SEID_ResultList)]
+        [DataRow(VSConstants.VSSELELEMID.SEID_StartupProject)]
+        [DataRow(VSConstants.VSSELELEMID.SEID_UndoManager)]
+        [DataRow(VSConstants.VSSELELEMID.SEID_UserContext)]
+        [DataRow(VSConstants.VSSELELEMID.SEID_WindowFrame)]
+        public void OnElementValueChanged_NonDocumentElementId_DoesNotRaiseEvent(VSConstants.VSSELELEMID elementId)
         {
-            var selectedFrame = CreateMockFrame(__WindowFrameTypeFlags.WINDOWFRAMETYPE_Document);
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS.UnitTests/ActiveDocumentTrackerTests.cs:132]: We no longer need the tests that create frames with different kinds of "frame type".

File: src/Infrastructure.VS.UnitTests/ActiveDocumentTrackerTests.cs [hunk 4]
```diff
@@ -114,115 +116,85 @@ public void OnSelectionChanged_DoesNotRaiseEvent()
...
+            var result = SimulateElementValueChanged(elementId, selectedFrame);
             result.Should().Be(VSConstants.S_OK);
 
             mockEventHandler.VerifyNoOtherCalls();
             textDocumentProviderMock.VerifyNoOtherCalls();
         }
 
         [TestMethod]
-        public void OnElementValueChanged_FrameElement_ToolFrame_DoesNotRaiseEvent()
+        public void OnElementValueChanged_DocumentElement_NewValueIsNull_RaisesEventWithNullArg()
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS.UnitTests/ActiveDocumentTrackerTests.cs:141]: We should now raise the event with a null arg. It means the last doc has been closed, or the doc isn't a text doc (e.g. is a graphical doc).

**Comment by duncanp-sonar** [comment]
FYI I raised another related "bug" in the same area: #2208. We can decide separately whether to fix it (the behaviour described in the new bug existed before and isn't changed by this fix).

**Review by rita-g-sonarsource — APPROVED** [review]

**Pull request merged**